app/models/concerns/annotating.rb

```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+module Annotating
+  extend ActiveSupport::Concern
+
+  class_methods do
+    def annotated
+      class_eval do
+        include Model
+        include ManagedBy
+      end
+    end
+  end
+
+  class << self
+    def models
+      return @models if @models
+
+      # this is to see all models in dev env, otherwise some may not be yet loaded
+      Rails.autoloaders.main.eager_load_dir("#{Rails.root}/app/models")
+
+      @models = ActiveRecord::Base.descendants.select { |model| model.include?(Model) }
+    end
+  end
+
+  module Model
+    extend ActiveSupport::Concern
+
+    included do
+      has_many :annotations, as: :annotated, dependent: :destroy, autosave: true
+    end
+
+    def annotation(name)
+      annotations.find { _1.name == name }
+    end
+
+    def value_of_annotation(name)
+      annotation(name)&.value
+    end
+
+    def annotate(name, value)
+      return remove_annotation(name) if value.nil?
+
+      existing = annotation(name)
+      if existing
+        existing.value = value
+      else
+        annotations.build(name: name, value: value, tenant_id: tenant_id)
```

Comment:
It was originally considered a security feature to separate tenants on the database level to prevent broken app logic from setting the wrong tenant. Something like a second layer of protection because app should already filter only objects from the proper tenant using other query criteria. And it was considered never or almost never `tenant_id` to be part of app payload to the database.

Reply:
Then you don't need to provide the `tenant_id` here because it's going to be added by the trigger.